src/MongoDB.AspNetCore.OData/MongoEnableQueryAttribute.cs

```diff
@@ -139,25 +148,41 @@ private static IQueryable ApplyProjection<T>(IMongoQueryable<T> queryable, OData
         return queryable.AppendStage(projectionStage);
     }
 
-    private static IQueryable ApplySelectExpand<T>(IMongoQueryable<T> queryable, ODataQueryOptions queryOptions)
+    private static IQueryable ApplyPaging<T>(IQueryable<T> queryable, int pageSize)
+        => new TruncatedCollection<T>(queryable, pageSize).AsQueryable();
+
+    private static IQueryable ApplySelectExpand<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
     {
         // SelectExpand projection generated by oData library is not compatible with Mongo LINQ provider,
         // so we split the query to server and client side queries, where SelectExpand with be executed on the client side.
-        var wrappedQueryable = queryable.ToEnumerable().AsQueryable();
+        var wrappedQueryable = queryable.ToArray().AsQueryable();
 
         var allowedOptions = AllowedQueryOptions.Select | AllowedQueryOptions.Expand | AllowedQueryOptions.Compute;
         var ignoreOption = AllowedQueryOptions.All & ~allowedOptions;
         return queryOptions.ApplyTo(wrappedQueryable, ignoreOption);
     }
 
-    private static IQueryable ApplyTransformationMethod(MethodInfo method, IQueryable queryable, ODataQueryOptions queryOptions)
+    private static IQueryable ApplyTransformationMethod(MethodInfo method, params object[] parameters)
     {
-        if (queryable is not IMongoQueryable)
+        var queryable = parameters[0];
+        if (!(queryable is IQueryable mongoQueryable))
         {
-            throw new ArgumentException("IMongoQueryable is expected.", nameof(queryable));
+            throw new ArgumentException("IQueryable is expected.", nameof(queryable));
         }
 
-        var genericMethod = method.MakeGenericMethod(queryable.ElementType);
-        return (IQueryable)genericMethod.Invoke(null, new object[] { queryable, queryOptions });
+        var genericMethod = method.MakeGenericMethod(mongoQueryable.ElementType);
+        return (IQueryable)genericMethod.Invoke(null, parameters);
     }
+
+    private ODataQuerySettings ToQuerySettings(TimeZoneInfo timeZoneInfo, AllowedQueryOptions ignoredQueryOptions)
+        => new ODataQuerySettings
+        {
+            TimeZone = timeZoneInfo,
+            EnsureStableOrdering = EnsureStableOrdering,
+            EnableConstantParameterization = EnableConstantParameterization,
+            HandleNullPropagation = HandleNullPropagation,
+            HandleReferenceNavigationPropertyExpandFilter = HandleReferenceNavigationPropertyExpandFilter,
+            EnableCorrelatedSubqueryBuffering = EnableCorrelatedSubqueryBuffering,
```

Comment:
Nope. This is copying values from `MongoEnableQueryAttribute` to newly created `ODataQuerySettings`. The confusion is because there are exactly the same property names.